go/marshal/decode.go

```diff
@@ -144,11 +155,28 @@ func uintDecoder(v types.Value, rv reflect.Value) {
 	}
 }
 
-var structDecoderCache struct {
+type decoderCacheT struct {
 	sync.RWMutex
```

Comment:
fwiw I bet that a mutex is as good as an rwlock here (and below). rwlocks typically make a bigger impact around longer hold times; otherwise the RFO of the cache line is going to dominate